subsys/bluetooth/audio/bap_unicast_server.c

```diff
@@ -28,10 +29,37 @@ LOG_MODULE_REGISTER(bt_bap_unicast_server, CONFIG_BT_BAP_UNICAST_SERVER_LOG_LEVE
 
 static const struct bt_bap_unicast_server_cb *unicast_server_cb;
 
+static struct bt_ascs_register_param ascs_param = {
+	CONFIG_BT_ASCS_ASE_SNK_COUNT,
+	CONFIG_BT_ASCS_ASE_SRC_COUNT
+};
+
+int bt_bap_unicast_server_set_ase_count(enum bt_audio_dir dir, uint8_t count)
+{
+	if (dir == BT_AUDIO_DIR_SINK) {
+		if (count > CONFIG_BT_ASCS_ASE_SNK_COUNT) {
+			return -EINVAL;
+		}
+		ascs_param.sink_ase_count = count;
+	} else if (dir == BT_AUDIO_DIR_SOURCE) {
+		if (count > CONFIG_BT_ASCS_ASE_SRC_COUNT) {
+			return -EINVAL;
+		}
+		ascs_param.source_ase_count = count;
+	}
+
+	return 0;
+}
+
 int bt_bap_unicast_server_register_cb(const struct bt_bap_unicast_server_cb *cb)
 {
 	int err;
 
+	err = bt_ascs_service_register(&ascs_param);
+	if (err != 0) {
+		return err;
+	}
```

Comment:
I do not like this. A function called `bt_bap_unicast_server_register_cb` should not affect the GATT database.

I suggest to add a new BAP unicast server function to trigger the ASCS registering

Reply:
@Thalley also linking your next comment here as they are quite related:
> Did you consider the approach of implementing two functions:
> 
> bt_bap_unicast_server_register(uint8_t src_cnt, uint8_t snk_cnt) that registers ASCS with the provided number of ASEs
> bt_bap_unicast_server_unregister that unregisters ASCS
> 
> That way it's very clear what the functions do, and we do not need to introduce a dependency on bt_bap_unicast_server_register_cb
 
That is fair enough, and I definetly see your point. The reason it was added here is that 'bt_bap_unicast_server_register_cb' initializes ascs, making ascs unusable unless you actually call cb registration first (wich makes sense, since some state machine transition needs input from user anyway).

Adding the gatt db registration here allows the dynamic ase configuraiton to be optional, and won't break anything already running.

For sure I can make the change you suggested - however, the Kconfigs then becomes a little confusing, and should maybe be renamed to e.g. 'CONFIG_BT_ASCS_MAX_ASE_SNK_COUNT' as they would never be used directly anymore.

Reply:
> Adding the gatt db registration here allows the dynamic ase configuraiton to be optional, and won't break anything already running.

That is a fair point. So it was a way to avoid API breakage? 

Since we are still in an unstable API I think it's better to break the API while we more easily can to add proper solutions now. 

With that in mind, we should also update the migration guide if we are breaking the API

Reply:
> That is a fair point. So it was a way to avoid API breakage?

Yes, this is what itched me the most.

>Since we are still in an unstable API I think it's better to break the API while we more easily can to add proper >solutions now.

This is fair enough. Ill make the change as suggested.

> With that in mind, we should also update the migration guide if we are breaking the API

Yes, I was just thinking about that :) Ill look into this